solidity/test/bridge/Bridge.MovingFunds.test.ts

```diff
@@ -0,0 +1,1005 @@
+import chai, { expect } from "chai"
+import { smock } from "@defi-wonderland/smock"
+import type { FakeContract } from "@defi-wonderland/smock"
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import { ContractTransaction } from "ethers"
+import type {
+  Bank,
+  BankStub,
+  Bridge,
+  BridgeStub,
+  BridgeStub__factory,
+  IRelay,
+  IWalletRegistry,
+} from "../../typechain"
+import bridgeFixture from "./bridge-fixture"
+import { walletState } from "../fixtures"
+import {
+  MovingFundsTestData,
+  MultipleInputs,
+  MultipleTargetWalletsAndDivisibleAmount,
+  MultipleTargetWalletsAndIndivisibleAmount,
+  MultipleTargetWalletsButAmountDistributedUnevenly,
+  SingleProvablyUnspendable,
+  SingleTargetWallet,
+  SingleTargetWalletButP2SH,
+} from "../data/moving-funds"
+
+chai.use(smock.matchers)
+
+const { createSnapshot, restoreSnapshot } = helpers.snapshot
+const { lastBlockTime } = helpers.time
+
+const fixture = async () => bridgeFixture()
+
+describe("Bridge - Moving funds", () => {
+  let treasury: SignerWithAddress
+
+  let bank: Bank & BankStub
+  let relay: FakeContract<IRelay>
+  let walletRegistry: FakeContract<IWalletRegistry>
+  let Bridge: BridgeStub__factory
+  let bridge: Bridge & BridgeStub
+
+  before(async () => {
+    // eslint-disable-next-line @typescript-eslint/no-extra-semi
+    ;({ treasury, bank, relay, walletRegistry, Bridge, bridge } =
+      await waffle.loadFixture(fixture))
+  })
+
+  describe("submitMovingFundsProof", () => {
+    context("when transaction proof is valid", () => {
+      context("when there is a main UTXO for the given wallet", () => {
+        context("when main UTXO data are valid", () => {
+          context("when there is only one input", () => {
+            context(
+              "when the single input points to the wallet's main UTXO",
+              () => {
+                context(
+                  "when the output vector references only 20-byte hashes",
+                  () => {
+                    context(
+                      "when the output vector has only P2PKH and P2WPKH outputs",
+                      () => {
+                        context(
+                          "when transaction amount is distributed evenly",
+                          () => {
+                            context(
+                              "when transaction fee is not too high",
+                              () => {
+                                context(
+                                  "when source wallet is in the MovingFunds state",
+                                  () => {
+                                    context(
+                                      "when target wallets commitment is submitted",
+                                      () => {
+                                        context(
+                                          "when actual target wallets correspond to the commitment",
+                                          () => {
```

Comment:
Non-blocking this PR but I think we could un-fold some of a deeply nested structures in unit tests, just like this one.

For example, it does not really matter if we ensure single input points to the wallet's main UTXO if the output vector has P2SH output.

We could first test unhappy paths on the same level of "nesting" and then have a happy path scenario. Something like:
```
when there is no main UTXO for the given wallet
  should revert
when main UTXO data are invalid
  should revert
when input count is other than one
  should revert
when the single input doesn't point to the wallet's main UTXO
  should revert
when the output vector doesn't only reference 20-byte hashes
  should revert
when transaction amount is not distributed evenly
  should revert
when transaction fee is too high
  should revert
when all input data are valid
  when there is a single target wallet
    should mark the main UTXO as correctly spent
    should unset the main UTXO for the source wallet
    should put the source wallet in the Closed state
(...)
```

I know we've been nesting quite heavily in this repo but this example is the most extreme one given it operates on a quite complex validation path.

Anyway, not blocking, we'll have a chance to revisit it when working on target wallets commitment function.

Reply:
I don't think nesting is something bad. We were writing our tests in different ways across Solidity repositories and, in my opinion, nesting works very well because it provides a good specification of how the code is supposed to work. It also allows introducing changes in an easy way. The biggest problem I see with the approach you mentioned is that it doesn't show all the dependencies. For example, in the scenario `when the transaction amount is not distributed evenly` the input vector and a lot of other things must be valid to test this scenario. However, the test structure does not make it clear. Also, we must initialize such preconditions for all cases separately. Now, if we use nesting, you can do initialization once at the appropriate level, and reuse it for dependent cases (a good example is here: https://github.com/keep-network/tbtc-v2/blob/749b9391b64c3a3fcc1eb2a1412a5f5620189e34/solidity/test/bridge/Bridge.Wallets.test.ts#L1075).

Reply:
I am not saying nesting is bad in general. The example you linked is a good use of nesting in my opinion. I just think that in the particular case of `Bridge.MovingFunds.test.ts` we could be nested too deeply, especially given that we are not doing any setup between nested contexts. Just something to think about.